### ISSUE TITLE
Bug: [no-unused-vars] False positive when returning an assignment

### ISSUE THREAD

**Issue by Mitsunee** [typescript-eslint/typescript-eslint#11782]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.9.2&fileType=.tsx&code=FAMwrgdgxgLglgewgAgPogQgRgQwE4AUAlMgN7DLIA2ApjMhggFxnK54taa04oC%2BAbmAVkcEMgIBZHDAAWAOjy8AJggC2xEuUqU8dMHhQFGyALyt2LGHjA1kfIkMp9hu-YaF8gA&eslintrc=N4KABGBEBOCuA2BTAzpAXGYBfEWg&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
function _foobar() {
  let foo: { bar: boolean };

  if (Math.random()) {
    return (foo = { bar: true });
  }

  return;
}
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-unused-vars": "warn",
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "moduleResolution": "Bundler",
    "module": "esnext",
    "target": "es2024",
    "lib": ["ES2024", "DOM"],
    "strict": true,
    "noEmit": true,
    "isolatedModules": true,
    "skipLibCheck": true
  },
  "include": ["**/*.ts"],
  "exclude": ["node_modules", "dist"]
}
```

### Expected Result

No warning should be given

### Actual Result

A warning is given despite the result of the assignment being used as the return value of the function in the same line

### Additional Info

This might seem a bit silly in isolation, but was simplified from a larger function where I created a `result` variable at the top to return various different results that should all match a given type interface to make narrowing of the result easier outside of the function.

**Comment by bradzacher** [comment]
This behaviour is consistent with the base rule [our playground with both rules turned on](https://typescript-eslint.io/play/#ts=5.9.2&fileType=.tsx&code=FAUwHgDg9gTgLgAgGYFcB2BjOBLKaED6SUUARgIYwAUAlAgN7AIIA2IixUAXAwhTD1Ik25fAF8A3MCYJsSBFQCy5OAAsAdDFEATKAFtadRs2Yx2KGPiqcEAXl78ecGChAIxNKczHTT5y1JiQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQHYHsBaWXWZRAE0IDcBDaVDSRaafaSAGnG0gAEALgE8ADigDG0AJYiBhFPCm4BAegLFS5KnQbooLNhx4BfEMaA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false) and [eslint playground with their rule turned on](https://eslint.org/play/#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLXVudXNlZC12YXJzOiBbXCJlcnJvclwiXSAqL1xuXG5leHBvcnQgZnVuY3Rpb24gX2Zvb2JhcigpIHtcbiAgbGV0IGZvbztcblxuICBpZiAoTWF0aC5yYW5kb20oKSkge1xuICAgIHJldHVybiAoZm9vID0geyBiYXI6IHRydWUgfSk7XG4gIH1cblxuICByZXR1cm47XG59XG4iLCJvcHRpb25zIjp7InJ1bGVzIjp7fSwibGFuZ3VhZ2VPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjoibGF0ZXN0Iiwic291cmNlVHlwZSI6Im1vZHVsZSIsInBhcnNlck9wdGlvbnMiOnsiZWNtYUZlYXR1cmVzIjp7fX19fX0=).

We seek to align with their reports; so we won't fix this until they do.
First thing to do is file an issue with them upstream and then if/when that gets accepted and fixed, we can fix it in our implementation too.

**Comment by Mitsunee** [comment]
As discussed in https://github.com/eslint/eslint/issues/20353 I'm closing this issue as well:

> I think this comes down to a misunderstanding of mine of how the `return (foo = value)` syntax works. The variable is merely assigned and the assigned value is *also* `return`ed, leaving the variable not used, only assigned, so actually this isn't a false positive at all and also doesn't achieve the TS behaviour I wanted, so here it *would* be better to use an explicit return type, which I try to avoid as it can lead to "lying" to TypeScript.

For anyone who comes across this issue in the future: You can use `return value satisfies Type` instead of what I was doing with a useless variable assignment. It'll have the same result, while still avoiding an explicit return type.
